### DISCUSSION TITLE
Backlash compensation logic.

### DISCUSSION THREAD

**Discussion by vegetate7** [kachurovskiy/nanoels#162] [Q&A]
Hello! This is completely still unclear for me. 
Let's start with Async code, because it's simple.
in the 'onAsyncTimer()':
```
  } else if (dupr > 0 && a->pos < a->leftStop) {
    if (a->pos >= a->motorPos) {
      a->pos++;
    }
    a->motorPos++;
    a->posGlobal++;
  } else if (dupr < 0 && a->pos > a->rightStop) {
    if (a->pos >= a->motorPos + a->backlashSteps) {
      a->pos--;
    }
    a->motorPos--;
    a->posGlobal--;

```
Let's assume we just powered it on, tool somewhere in the middle and tool actual position zeroed, 0.
So:
posGlobal==0, motorPos==0, originPos == 0, pos == 0
therefore display position == 0 (pos+originPos)
set leftStop = 100, rightStop = -100
backlashSteps, let's take for simple 10. And toll are in the middle of backlash (between threads). 5 steps to both directions.
Starting Async mode.
Cycle 1:
dupr positive.
because pos == motorPos, both incrementing simultaneous, motor rotating left, taking 5 steps of backlash, then tool start moving. pos going to 100, we stops after 100 steps.
pos == motorPos == 100, display == 100, actual tool position 95 (remember backlash?)
Cycle 2:
swith dupr to negative.
because pos == motorPos, decremenitg motorPos alone for backlashSteps, motor rotating right, taking 10 steps of backlash, then tool start moving, and pos start decrementing. pos going to -100, we stops after 210 steps.
Now, pos == -100, motorPos == -110, display == -100. actual tool position (95 - (210-10)) = -105.
Ok, Cycle 3:
swith dupr to positive.
because pos >= motorPos, both incrementing simultaneous, motor rotating left, taking 10 steps of backlash, then tool start moving. pos going to 100, we stops after 200 steps.
And now pos == 100, motorPos = 90, display == 100. Actual tool position (-105 + (200 - 10)) = 85! We are actually drifted 10 steps right!
Ok, I think it's bug in the code, insted of
```
   if (a->pos >= a->motorPos) {
      a->pos++;
   }


```
should be
```
 if (a->pos <= a->motorPos) {
      a->pos++;
 }

```
Or use abs() may be. It should stop drifting.
But I afraid it's just the tip of big pile of backlash problems.

PS: sorry for multiple edits, that text formatting killing me.

**Comment by vegetate7** [comment]
Or, may be it's just a bug in AsyncTimer(), and may be not all so bad :)
This part in 'moveAxis()' looks good.
So, may be, the problem lasts is the position shift, produced by starting in the middle of leadscrew thread. But it's small and it's constant, not driftint to any direction.

**Comment by kachurovskiy** [comment]
Nice find :) Yes, looks like it should be as you suggested. https://github.com/kachurovskiy/nanoels/commit/db2f4a05efdf58fd0f9e65be44d7b3d9adc0b703
